### DISCUSSION TITLE
VCI Directory Snapshot Creation

### DISCUSSION THREAD

**Discussion by christianpaquin** [christianpaquin/vci-directory-auditor#3] [Ideas]
# VCI Directory Snapshot Creation

It is desirable to frequently create a snapshot of the VCI directory, including the issuers’ keys and revocation information (e.g., CRLs), and make it available to verifiers. This is motivated by the fact that some verifiers might want to operate offline, might want a back-up in case of connectivity issues, or are unable to access some issuer endpoints due to geolocation issues.

This proposal recommends that:

1.	VCI creates and publishes a script to assemble a snapshot of the directory
2.	VCI runs said script to package a directory daily, using a set of leniency rules to copy unreachable/erroneous issuer’s information from a previous snapshot
3.	VCI packages the snapshot into a signed bundle for redistribution

## Notes
* The [auditing scripts](https://github.com/the-commons-project/vci-directory/tree/main/scripts/vci-directory-auditor) part of the VCI github repository contains code to create a snapshot, but it should be extracted into its own independent script to allow verifiers to easily run it themselves if desired.

* Proposed packaging rules are:
   1. Each listed issuer's JWKS (and CRLs, if any) is retrieved daily, errors are noted in the `vci_log.json` file.
   1. The daily snapshot is generated from all retrieved data, combined with the missing content from the previous X days (recommended X = 3). This allows unavailable issuers to remain in the daily snapshot if they experience technical issues.
   1. JWKS of issuer that are unavailable for more than X days are added to an archival snapshot. This will allow verifiers to validate valid SHCs from non-active issuers.
   1. The `kid` of revoked issuer keys (keys removed from a JWKS) are added to a revoked snapshot. This will allow verifiers to treat unknown and revoked keys differently (e.g., in error messages)
   1. The resulting `vci_snapshot.json`, `vci_archive.json`, `vci_revoked.json`, and the `vci_log.json` are committed to github.
   1. The files are signed with a VCI signing key, the public key of which is available on the VCI website, the file signatures are also committed to github.

* The resulting snapshot files should be made available on a non-github location, e.g., on VCI.org.

**Comment by jmandel** [comment]
Is `vci_archive.json` only for things that are unavailable, or does it contain full logs daily?

Is there a predictable way to get historical state (e.g., git tags or something on a daily basis) or do the JSON files include full historical state in each commit?

> The daily snapshot is generated from all retrieved data, combined with the missing content from the previous X days

Is there an indication of which data have been "filled in" vs directly observed? Or do you check the archive file for "directly observed" data (see question above on archive)?

**Reply by christianpaquin** [reply]
> Is `vci_archive.json` only for things that are unavailable, or does it contain full logs daily?

That would be a special file for keys that once were in the VCI directory, but are not available anymore (e.g., a pharmacy went under, or a medical org lost access to a web endpoint (re-org, acquisition, etc.)). Keys that are not part of the daily snapshot, and have not been ruled out due to some rule violations could be added to the archive.

> Is there a predictable way to get historical state (e.g., git tags or something on a daily basis) or do the JSON files include full historical state in each commit?

My prototype overwrites the snapshot every day, and you use github history to fetch the version you want. We could output the daily commit tag somewhere, or use different names everyday `log-<date>.json`

> Is there an indication of which data have been "filled in" vs directly observed? Or do you check the archive file for "directly observed" data (see question above on archive)?

Question is: would validation app make a distinction (or be forced to?) An auditor has the data to figure this out, but I feel like verifiers just want "the list", offloading the decision making to VCI (and they rely on out snapshot vs. creating it themselves). OTOH, we can certainly make a label (or last-fetched-timestamp) for each entry.

**Comment by jmandel** [comment]
I think it'd also be interesting to create an (append-only) sqlite db file with full requests and responses from each day's run, and generate other artifacts from that db.

**Reply by christianpaquin** [reply]
Good to discuss the ideas, then we can turn that into a spec, and start building the pieces.

**Comment by jmandel** [comment]
>An auditor has the data to figure this out, but I feel like verifiers just want "the list", offloading the decision making to VCI

This makes sense to me; I just wasn't sure where "the data to figure it out" would live among the three files as currently documented. If it's not in these three files, are there other outputs that are also saved and made available for auditors?

**Reply by christianpaquin** [reply]
Yes, was missing one file; just updated the text above to contain the `vci_log.json`, which would be a raw version of the retrieval (keys, errors, etc.), from which the other app-queryable files would be created. (That's already implemented in VCI's [PR #454](https://github.com/the-commons-project/vci-directory/pull/454))
